internal/dbunique/db_unique.go

```diff
@@ -32,195 +28,127 @@ var defaultUniqueStates = []rivertype.JobState{ //nolint:gochecknoglobals
-
-	advisoryLockHash := hashutil.NewAdvisoryLockHash(i.AdvisoryLockPrefix)
-	advisoryLockHash.Write([]byte("unique_key"))
-	advisoryLockHash.Write([]byte("kind=" + params.Kind))
-	advisoryLockHash.Write([]byte(uniqueKey))
-
-	getParams := i.buildGetParams(params, uniqueOpts)
-
-	// Begin a subtransaction
-	subExec, err := exec.Begin(ctx)
-	if err != nil {
-		return nil, err
-	}
-	defer subExec.Rollback(ctx)
-
-	// The wrapping transaction should maintain snapshot consistency even if we
-	// were to only have a SELECT + INSERT, but given that a conflict is
-	// possible, obtain an advisory lock based on the parameters of the unique
-	// job first, and have contending inserts wait for it. This is a synchronous
-	// lock so we rely on context timeout in case something goes wrong and it's
-	// blocking for too long.
-	if _, err := subExec.PGAdvisoryXactLock(ctx, advisoryLockHash.Key()); err != nil {
-		return nil, fmt.Errorf("error acquiring unique lock: %w", err)
-	}
-
-	existing, err := subExec.JobGetByKindAndUniqueProperties(ctx, getParams)
-	if err != nil {
-		if !errors.Is(err, rivertype.ErrNotFound) {
-			return nil, fmt.Errorf("error getting unique job: %w", err)
-		}
-	}
-
-	if existing != nil {
-		// Insert skipped; returns an existing row.
-		return &rivertype.JobInsertResult{Job: existing, UniqueSkippedAsDuplicate: true}, nil
-	}
-
-	jobRow, err := subExec.JobInsertFast(ctx, params)
-	if err != nil {
-		return nil, err
-	}
-
-	if err := subExec.Commit(ctx); err != nil {
-		return nil, err
+func (o *UniqueOpts) StateBitmask() byte {
+	states := defaultUniqueStates
+	if len(o.ByState) > 0 {
+		states = o.ByState
 	}
+	return UniqueStatesToBitmask(states)
+}
 
-	return &rivertype.JobInsertResult{Job: jobRow}, nil
+func UniqueKey(timeGen baseservice.TimeGenerator, uniqueOpts *UniqueOpts, params *riverdriver.JobInsertFastParams) []byte {
+	uniqueKeyString := buildUniqueKeyString(timeGen, uniqueOpts, params)
+	uniqueKeyHash := sha256.Sum256([]byte(uniqueKeyString))
+	return uniqueKeyHash[:]
 }
 
 // Builds a unique key made up of the unique options in place. The key is hashed
 // to become a value for `unique_key` in the fast insertion path, or hashed and
 // used for an advisory lock on the slow insertion path.
-func (i *UniqueInserter) buildUniqueKey(params *riverdriver.JobInsertFastParams, uniqueOpts *UniqueOpts) (string, bool) {
+func buildUniqueKeyString(timeGen baseservice.TimeGenerator, uniqueOpts *UniqueOpts, params *riverdriver.JobInsertFastParams) string {
 	var sb strings.Builder
 
+	if !uniqueOpts.ExcludeKind {
+		sb.WriteString("&kind=" + params.Kind)
+	}
+
 	if uniqueOpts.ByArgs {
+		encodedArgsForUnique := params.EncodedArgs
+
+		{
+			// Get unique JSON keys from the JobArgs struct:
+			uniqueFields, err := getSortedUniqueFieldsCached(params.Args)
+			if err != nil {
+				// Instead of returning an error, we can just proceed using the entire
+				// encoded args.
+				//
+				// TODO: log or how will we find out about this?
```

Comment:
This is one of those "should never happen" errors. I believe it can only happen if you give us a non-struct `JobArgs`. Is that even possible or does it happen in the wild? 🤷‍♂️